cpu/atmega_common/periph/gpio.c

```diff
@@ -59,7 +63,101 @@
 #endif
 
 static gpio_isr_ctx_t config[GPIO_EXT_INT_NUMOF];
-#endif /* MODULE_PERIPH_GPIO_IRQ */
+
+/**
+ * @brief detects ammount of possible PCINTs
+ */
+#if defined(MODULE_ATMEGA_PCINT0) || defined(MODULE_ATMEGA_PCINT1) || defined(MODULE_ATMEGA_PCINT2) || defined(MODULE_ATMEGA_PCINT3)
+#include "atmega_pcint.h"
+
+/**
+ * @brief check which pcints should be enabled!
+ */
+#ifdef MODULE_ATMEGA_PCINT0
+#ifndef ATMEGA_PCINT_MAP_PCINT0
+#error Either mapping for pin change interrupt bank 0 is missing or not supported by the MCU
+#else
+#define PCINT0_IDX (0)
+#define _COUNTER0  (1)
+#endif /* ATMEGA_PCINT_MAP_PCINT0 */
+#else
+#define _COUNTER0  (0)
+#endif /* MODULE_ATMEGA_PCINT0 */ 
+
+#ifdef MODULE_ATMEGA_PCINT1
+#ifndef ATMEGA_PCINT_MAP_PCINT1
+#error Either mapping for pin change interrupt bank 1 is missing or not supported by the MCU
+#else
+#define PCINT1_IDX _COUNTER0
+#define _COUNTER1 (_COUNTER0 + 1)
+#endif /* ATMEGA_PCINT_MAP_PCINT1 */
+#else
+#define _COUNTER1 _COUNTER0
+#endif /* MODULE_ATMEGA_PCINT1 */ 
+
+#ifdef MODULE_ATMEGA_PCINT2
+#ifndef ATMEGA_PCINT_MAP_PCINT2
+#error Either mapping for pin change interrupt bank 2 is missing or not supported by the MCU
+#else
+#define PCINT2_IDX _COUNTER1
+#define _COUNTER2 (_COUNTER1 + 1)
+#endif /* ATMEGA_PCINT_MAP_PCINT2 */
+#else
+#define _COUNTER2 _COUNTER1
+#endif /* MODULE_ATMEGA_PCINT2 */ 
+
+#ifdef MODULE_ATMEGA_PCINT3
+#ifndef ATMEGA_PCINT_MAP_PCINT3
+#error Either mapping for pin change interrupt bank 3 is missing or not supported by the MCU
+#else
+#define PCINT3_IDX _COUNTER2
+#define _COUNTER3 (_COUNTER2 + 1)
+#endif /* ATMEGA_PCINT_MAP_PCINT3 */
+#else
+#define _COUNTER3 _COUNTER2
+#endif /* MODULE_ATMEGA_PCINT3 */ 
+
+#define PCINT_NUM_BANKS (_COUNTER3)
+
+/**
+ * @brief stores the last pcint state of each port
+ */
+static uint8_t pcint_state[ PCINT_NUM_BANKS ];
+
+/**
+ * @brief stores all cb and args for all defined pcint.
+ */
+typedef struct {
+    gpio_cb_t cb;           /**< interrupt callback */
+    gpio_t pin;             /**< pin which the interrupt is for */
+    gpio_flank_t flank;     /**< type of interrupt the flank should be triggered on */
+    void *arg;              /**< optional argument */
+} gpio_isr_ctx_pcint_t;
+
+/**
+ * @brief
+ */
+static const gpio_t pcint_mapping[][8] = {
```

Comment:
Might be possible, does it really do much? Isn't it just a notation thing? Memory-wise it should be exactly the same?

Reply:
By being able to have a single loop iterating over it instead of two nested loops I'd bet a few instructions are saved. The array itself will have the same size regardless of being 1d or 2d.